requests/signed_message.go

```diff
@@ -0,0 +1,160 @@
+package requests
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// SignedRequest is a builder for constructing a http.Request with Yoti signing
+type SignedRequest struct {
+	Key        *rsa.PrivateKey
+	HTTPMethod string
+	BaseURL    string
+	Endpoint   string
+	Headers    map[string][]string
+	Params     map[string]string
+	Body       []byte
+	Error      error
+}
+
+func (msg *SignedRequest) signDigest(digest []byte) (string, error) {
+	hash := sha256.Sum256(digest)
+	signed, err := rsa.SignPKCS1v15(rand.Reader, msg.Key, crypto.SHA256, hash[:])
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(signed), nil
+}
+
+func getTimestamp() string {
+	return strconv.FormatInt(time.Now().Unix()*1000, 10)
+}
+
+func getNonce() (string, error) {
+	nonce := make([]byte, 16)
+	_, err := rand.Read(nonce)
+	return fmt.Sprintf("%X-%X-%X-%X-%X", nonce[0:4], nonce[4:6], nonce[6:8], nonce[8:10], nonce[10:]), err
+}
+
+// WithPemFile loads the private key from a PEM file reader
+func (msg SignedRequest) WithPemFile(in []byte) SignedRequest {
+	block, _ := pem.Decode(in)
+	if block == nil {
+		msg.Error = errors.New("Not PEM-encoded")
+		return msg
+	}
+	if block.Type != "RSA PRIVATE KEY" {
+		msg.Error = errors.New("Not an ESA Private Key")
+		return msg
+	}
+
+	msg.Key, msg.Error = x509.ParsePKCS1PrivateKey(block.Bytes)
+	return msg
+}
+
+// Request builds a http.Request with signature headers
+func (msg SignedRequest) Request() (request *http.Request, err error) {
+	if msg.Error != nil {
+		return nil, msg.Error
+	}
+	// Check for mandatorys
+	if msg.Key == nil {
+		err = fmt.Errorf("Missing Private Key")
+		return
+	}
+	if msg.HTTPMethod == "" {
+		err = fmt.Errorf("Missing HTTPMethod")
+		return
+	}
+	if msg.BaseURL == "" {
+		err = fmt.Errorf("Missing BaseURL")
+		return
+	}
+	if msg.Endpoint == "" {
+		err = fmt.Errorf("Missing Endpoint")
+		return
+	}
+
+	if msg.Params == nil {
+		msg.Params = make(map[string]string)
+	}
+	if _, ok := msg.Params["nonce"]; !ok {
+		nonce, err := getNonce()
+		if err != nil {
+			return nil, err
+		}
+		msg.Params["nonce"] = nonce
+	}
+	if _, ok := msg.Params["timestamp"]; !ok {
+		msg.Params["timestamp"] = getTimestamp()
+	}
+
+	// Add Timestamp/Nonce to Endpoint
+	endpoint := msg.Endpoint
+	if !strings.Contains(endpoint, "?") {
+		endpoint = endpoint + "?"
+	} else {
+		endpoint = endpoint + "&"
+	}
+
+	var firstParam = true
+	for param, value := range msg.Params {
+		var formatString = "%s&%s=%s"
+		if firstParam {
+			formatString = "%s%s=%s"
+		}
+		endpoint = fmt.Sprintf(formatString, endpoint, param, value)
+		firstParam = false
+	}
+
+	// Generate and Sign the message digest
```

Comment:
Since this function is quite long, would be good if we can split some of this functionality out. Signing the message digest would be a good one, maybe other parts too